File: python/taflow/indicators/rectangle_range.py
"""Python adapter for native rectangle-range detection."""

from typing import Any

import numpy as np

from .._native import RectangleRange as _Native
from .._series import as_float64_series


class RectangleRange:
    """Detect horizontal support-and-resistance rectangles from swing pivots.

    The native state confirms 5% swings and checks whether the last two highs
    and lows are each flat within 3%. A support touch emits ``1.0``, a resistance
    touch emits ``-1.0``, and all other bars emit ``0.0``. Open and close are
    accepted for the common OHLC contract but geometry uses high and low. This
    maps to Wickra ``RectangleRange``.

    Args:
        open: Required opening-price history.
        high: Required high-price history.
        low: Required low-price history.
        close: Required closing-price history.

    Raises:
        ValueError: If the four OHLC histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(
        self, open: float, high: float, low: float, close: float
    ) -> "RectangleRange":
        """Append one OHLC bar and return this adapter."""
        self._state.append(float(open), float(high), float(low), float(close))
        return self

    def extend(self, open: Any, high: Any, low: Any, close: Any) -> "RectangleRange":
        """Append aligned OHLC histories after validating their lengths."""
        arrays = tuple(as_float64_series(item) for item in (open, high, low, close))
        if len({len(item) for item in arrays}) != 1:
            raise ValueError("OHLC inputs must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest range signal, or ``None`` before the first bar."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned ``-1``, ``0``, and ``1`` signal history."""
        return self._state.compute()

    def reset(self) -> "RectangleRange":
        """Clear confirmed pivots and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-bar count delegated to native state."""
        return len(self._state)

File: crates/taflow-core/src/indicators/rectangle_range.rs
use crate::error::TaResult;

use super::pattern_swing::{
    approximately_equal, recent_legs, SwingTracker, LEVEL_TOLERANCE, SWING_THRESHOLD,
};

/// Causal horizontal support-and-resistance range detector.
#[derive(Debug, Clone)]
pub struct RectangleRange {
    swing: SwingTracker,
    count: usize,
    value: Option<f64>,
}

impl RectangleRange {
    /// Create a detector using Wickra's fixed swing and level tolerances.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            swing: SwingTracker::new(SWING_THRESHOLD, 4),
            count: 0,
            value: None,
        })
    }

    /// Append one OHLC bar and return the latest range-touch signal.
    pub fn append(&mut self, _open: f64, high: f64, low: f64, _close: f64) -> Option<f64> {
        self.count += 1;
        self.value = Some(0.0);
        if !self.swing.append(high, low) {
            return self.value;
        }
        let pivots = self.swing.pivots();
        if pivots.len() < 4 {
            return self.value;
        }

        let (old_high, new_high, old_low, new_low) = recent_legs(pivots);
        let flat_highs = approximately_equal(old_high, new_high, LEVEL_TOLERANCE);
        let flat_lows = approximately_equal(old_low, new_low, LEVEL_TOLERANCE);
        if flat_highs && flat_lows {
            let last_is_high = pivots[pivots.len() - 1].direction > 0.0;
            self.value = Some(if last_is_high { -1.0 } else { 1.0 });
        }
        self.value
    }

    /// Return the signal produced for the latest bar.
    pub fn value(&self) -> Option<f64> {
        self.value
    }

    /// Return the number of processed bars.
    pub fn len(&self) -> usize {
        self.count
    }

    /// Return whether no bars have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }

    /// Clear pivots and restore fresh-state behaviour.
    pub fn reset(&mut self) {
        self.swing.reset();
        self.count = 0;
        self.value = None;
    }
}
